runtime-sdk/src/module.rs

```diff
@@ -370,6 +373,54 @@ impl TransactionHandler for Tuple {
     }
 }
 
+/// Roothash incoming message handler.
+pub trait IncomingMessageHandler {
+    /// Add storage prefixes to prefetch, except for the prefixes for the embedded transaction. The
+    /// dispatcher will invoke the method handler for the embedded transaction separately.
+    fn prefetch_in_msg(
+        _prefixes: &mut BTreeSet<Prefix>,
+        _in_msg: &roothash::IncomingMessage,
+        _data: &IncomingMessageData,
+        _tx: &Option<Transaction>,
+    ) -> Result<(), error::RuntimeError> {
+        Ok(())
+    }
+
+    /// Execute an incoming message, except for the embedded transaction. The dispatcher will
+    /// invoke the transaction and method handlers for the embedded transaction separately.
```

Comment:
We should probably explicitly mention that returning an error here will cause a batch abort (and possibly stall the whole runtime as any future batch will also include this same message). So returning errors here must be considered carefully.

Also it should be noted that this executes in the batch-wide context and so any effects will not be rolled back even if the inner transaction fails.